### PR TITLE
vmm: Add support for explicit PCI device IDs

### PR DESCRIPTION
**Pull request by rbradford**
Based on the work in #7631 but adjusted to correctly handle clashes between
device types (not just within virtio devices). Building on the refactoring from
#7962

- **pci: Refactor bus.rs to better fit a PCI bus's semantics**
- **vmm: Allow for device ID allocation on a segment**
- **vmm: Fix segment log message formatting**
- **vmm: Add tests for `allocate_device_id` in `PciSegment`**
- **pci: Add support for reserving but not allocating slots**
- **vmm: Use the PciBus::allocate_device_id on restore path**
- **vmm: config: Add pci_device_id to PciDeviceCommonConfig**
- **vmm: openapi: Add pci_device_id to the required device entries**
- **vmm: config: Add pci_device_id to SYNTAX for supported devices**
- **docs: Update the relevant documentation**
- **vmm: Validate PCI device ID**
- **vmm: Propagate PCI device ID from the config**
- **vmm: device_manager: Reserve explicitly used PCI device IDs**
- **tests: Return stderr when executing commands**
- **tests: Add an integration test to verify PCI device allocations**
- **tests: Add an integration test to check duplicate PCI device IDs**
- **tests: Add an integration test for PCI device ID allocation errors**

### CODE REVIEW

**Comment by phip1611** [comment]
@scholzp please run our internal test suite (just the BDF test cases without live migration) against this. Does it work?

File: vmm/src/vm_config.rs [hunk 1]
```diff
@@ -263,15 +263,27 @@ pub struct VirtQueueAffinity {
...
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize, Default)]
+pub struct PciDeviceCommonConfig {
+    #[serde(default)]
+    pub id: Option<String>,
+    #[serde(default, skip_serializing_if = "<&bool as std::ops::Not>::not")]
+    pub iommu: bool,
+    #[serde(default)]
+    pub pci_segment: u16,
+    #[serde(default)]
+    pub pci_device_id: Option<u8>,
```
Comment (phip1611) [inline vmm/src/vm_config.rs:285]: For the record: I think this is a breaking change for us; we call the field `bdf_device`. This is not a major dealbreaker for us as we always ship `libvirt/ch` and `cloud-hypervisor` in matching pairs - just a heads-up.

Would it be possible that, for a grace period, we add  `#[serde(alias = "bdf_device")]`? This of course would require that @scholzp successfully runs this with the alias against our test suite (the BDF test cases)

Comment (scholzp) [reply vmm/src/vm_config.rs:285]: I applied the renaming together with `#[serde(alias = "bdf_device")]` to our version of the explicit device IDs (that of #7631). While instantiating a VM seems to work with this workaround, ~~some of our live migration tests fail. This indicates that there is some other side effect that I'm investigating right now. I'll come back to you during the day.~~

Edit: Further investigation showed that all is fine and we perhaps had some resource issues when testing the live migrations. @phip1611  and I also agreed on that we do not need to introduce `#[serde(alias = "bdf_device")]` here but rather resolve potential naming clashes downstream as long as we need to.

**Comment by rbradford** [comment]
@phip1611 @scholzp I've rebased

File: pci/src/bus.rs [hunk 2]
```diff
@@ -172,13 +179,37 @@ impl PciBus {
...
+                Err(PciRootError::AlreadyInUsePciDeviceSlot(idx))
+            }
+        } else {
+            Err(PciRootError::InvalidPciDeviceSlot(idx))
+        }
+    }
+
     /// Allocates a PCI device ID on the bus.
     ///
     /// - `id`: ID to allocate on the bus. If [`None`], the next free
```
Comment (phip1611) [inline pci/src/bus.rs:209]: nit: commit `pci: Add support for reserving but not allocating slots` needs adapted documentation

How about a short

```
- `id`: Preferred ID to reserve on the bus. If [`None`], the next free device ID will be used.`
```

File: vmm/src/vm_config.rs [hunk 3]
```diff
@@ -281,6 +281,8 @@ pub struct PciDeviceCommonConfig {
     pub iommu: bool,
     #[serde(default)]
     pub pci_segment: u16,
+    #[serde(default)]
+    pub pci_device_id: Option<u8>,
```
Comment (phip1611) [inline vmm/src/vm_config.rs:285]: This is an option to
- support migration from old to new, aye? Devices spawned with a new Cloud Hypervisor will never set this field to None?
- allow users to not pass a `pci_device_id` (the next free ID will be allocated)

yes?

File: vmm/src/config.rs [hunk 4]
```diff
@@ -402,6 +402,13 @@ pub enum ValidationError {
     /// Invalid NUMA Configuration
     #[error("NUMA Configuration is invalid")]
     InvalidNumaConfig(String),
+    /// The supplied PCI ID was greater then the max. supported number
+    /// of devices per Bus
+    #[error("Given PCI device ID ({0:02x}) is out of the supported range of 0..31")]
```
Comment (phip1611) [inline vmm/src/config.rs]: Isn't this `0..=31` / `0..32`? (31 is inclusive)?

File: vmm/src/config.rs [hunk 5]
```diff
@@ -414,6 +421,21 @@ pub fn add_to_config<T>(items: &mut Option<Vec<T>>, item: T) {
     }
 }
 
+// Check that the PCI device supplied is neither out of range nor does
```
Comment (phip1611) [inline vmm/src/config.rs]: nit: `//`--> `///` (rustdoc)

File: vmm/src/config.rs (hunk 4 above)
Comment (phip1611) [reply vmm/src/config.rs]: You could also use `0..{pci::NUM_DEVICE_IDS}` there I think.

**Review by phip1611 — COMMENTED** [review]
Lovely! Just a few nits.

PS: @scholzp will finish his final review also today before ~2pm - looking good!

To summarize for my own understanding:

**Phase 1** — `reserve_explicit_device_ids()` is called once before any device
creation. It iterates `virtio_devices`, `config.devices`, and
`config.user_devices`, and for every entry with `pci_device_id: Some(id)` calls
`PciBus::reserve_device_id(id)`: `Free → Reserved`. This blocks anonymous
allocation from stealing explicitly requested slots without yet committing them.

**Phase 2** — per-device, when actually added to the bus, `pci_resources()` calls
`PciSegment::allocate_device_id(pci_device_id)`:
- `Some(id)`: `Reserved → Allocated` (also accepts `Free`, rejects `Allocated`)
- `None`: scans for first `Free` slot, skipping `Reserved` and `Allocated`

Both virtio-net and vhost-user reach this via
`add_virtio_pci_device(..., pci_device_id)` → `pci_resources(..., pci_device_id)`;
VFIO/vfio-user call `pci_resources` directly.

State machine:

```
    Free ──reserve──► Reserved ──allocate(Some)──► Allocated
    Free ─────────────────────────allocate(None)──► Allocated
    Allocated/Reserved ──────put──────────────────► Free
```

Duplicate explicit IDs fail at phase 1; out-of-range and reserved IDs (e.g. the root device at slot 0) fail at config validation before either phase runs.

File: cloud-hypervisor/tests/integration.rs [hunk 6]
```diff
@@ -5834,6 +5834,84 @@ mod common_parallel {
 
         handle_child_output(r, &output);
     }
+
+    #[test]
+    // Test that adding a duplicate PCI device ID fails
+    fn test_duplicate_pci_device_id() {
```
Comment (scholzp) [inline cloud-hypervisor/tests/integration.rs:5850]: Very nit (feel free to ignore):  The commit message titled "tests: Add an integration test to check duplicate PCI device IDs" contains unnecessary white space (between twice and Moreover).

"
This integration test verifies that the same device ID cannot be
allocated <ins>twice.__Moreover</ins>, we check that the returned error matches our
expectations.
"

File: cloud-hypervisor/tests/integration.rs [hunk 7]
```diff
@@ -5912,6 +5912,86 @@ mod common_parallel {
 
         handle_child_output(r, &output);
     }
+
+    #[test]
+    // Test that requesting an invalid device ID fails.
+    fn test_invalid_pci_device_id() {
```
Comment (scholzp) [inline cloud-hypervisor/tests/integration.rs:5928]: The commit message titled "tests: Add an integration test for PCI device ID allocation errors" doesn't make sense anymore. It originally contained a second part referring to function IDs. This part is now missing and we should rewrite it to fit the new test. I though about something like:

"Check that CHV returns the correct error on allocation of an invalid device ID (one that is not in the range 0-31) and when trying to allocate a reserved ID (such as that of the root bridge)."

File: pci/src/bus.rs [hunk 8]
```diff
@@ -20,9 +20,13 @@ use crate::configuration::{
 };
 use crate::device::{BarReprogrammingParams, DeviceRelocation, Error as PciDeviceError, PciDevice};
 
+/// Denotes the PCI device ID of a bus' root bridge device.
+pub const PCI_ROOT_DEVICE_ID: u8 = 0;
+/// Denotes the maximum number of PCI device allowed on a bus. 32 per PCI spec.
```
Comment (scholzp) [inline pci/src/bus.rs]: Nit: I just found a typo that I introduced. Sadly, I cannot fix it (there is a `s` missing in `devices`).

```
/// Denotes the maximum number of PCI devices allowed on a bus. 32 per PCI spec.
```

File: vmm/src/vm_config.rs (hunk 3 above)
Comment (rbradford) [reply vmm/src/vm_config.rs:285]: Indeed both! And support migration of course.

File: vmm/src/config.rs (hunk 4 above)
Comment (rbradford) [reply vmm/src/config.rs]: Yes and in fact I think it's confusing it's in hex, but the value provide by the user would be decimal so I will fix that.

**Review by scholzp — APPROVED** [review]
I found a few typos, one of which I have authored. Otherwise I think all looks good.

Thanks again for investing your time! :)

File: pci/src/bus.rs (hunk 2 above)
Comment (rbradford) [reply pci/src/bus.rs:209]: Not an `Option` but, yeh, good spot. Done.

**Review by sboeuf — APPROVED** [review]
